### PR TITLE
feat: auto renew bot & indexer

### LINKED ISSUE #1: Automatic renewal bot
The goal of this bot is to detect expiring domain linked to "automatic renewal contract" and renew them automatically. We could have multiple implementations of it. Anyone should be able to run one.

### PR DESCRIPTION
**Pull request by irisdv**
Close #1 

This PR implements 2 crates :
- Indexer: index domain to addr update, addr to domain update, domain transfer, starknetid update, toggled renewals and approval events.
- Bot : renew bot which runs every 24 hours, fetches domains with an expiry date < 30 days, checks auto-renew has been enabled and that renewer has a sufficient allowance. 

**Testing :** 
You can test the indexer and renewal bot on the devnet.

If not on macos you can use `docker-compose` configuration file in this repo. 
Otherwise, directly run devnet using flags `--seed 0` and `--timeout 5000` and apibara using this command `OTEL_SDK_DISABLED=true RUST_LOG=info cargo run --release -p apibara-starknet -- start --devnet --rpc http://localhost:5050/rpc --wait-for-rpc`

Testing scripts are available in this PR : https://github.com/starknet-id/auto_renew_contract/pull/2

### CODE REVIEW

File: indexer/src/processing.rs [hunk 1]
```diff
@@ -13,63 +14,31 @@ use apibara_core::{
...
-                } => {
-                    if finality != DataFinality::DataStatusFinalized {
-                        println!("shutting down");
-                        break;
-                    }
-
-                    for block in batch {
-                        process_block(&conf, &state, block).await?;
-                    }
+    while let Some(message) = data_stream.try_next().await.unwrap() {
```
Comment (Th0rgal) [inline indexer/src/processing.rs:24]: please try to remove this unwrap, the connection will probably get closed at some point, you can check https://github.com/starknet-id/analysor

File: indexer/src/processing.rs [hunk 2]
```diff
@@ -13,63 +14,31 @@ use apibara_core::{
...
-            },
-            Ok(None) => {
-                break;
             }
-            Err(e) => {
-                eprintln!("Error while processing data stream: {:?}", e);
-                tokio::time::sleep(Duration::from_secs(5)).await;
-                continue;
+            DataMessage::Invalidate { cursor } => {
+                panic!("chain reorganization detected: {cursor:?}");
```
Comment (Th0rgal) [inline indexer/src/processing.rs:41]: An alert would be nice when the program panic, we certainly need to look at it

**Review by Th0rgal — COMMENTED** [review]
A few things to change. Rule of thumb for unwrap:
- It's a value we provide in the config: you can keep it
- It's a value that comes from a friendly actor (e.g. apibara): try to manage it if possible, but it's not a big deal
- It could fail for a random reason (a connection stopped) or worse if someone does it intentionally: we must take care of it

File: bot/src/config.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    contract: Contract,
+    database: Database,
+    devnet_provider: DevnetProvider,
+    account: MyAccount,
+    discord: Discord,
+});
+
+pub fn load() -> Config {
+    let args: Vec<String> = env::args().collect();
+    let config_path = if args.len() <= 1 {
```
[the remark this answers is not in the document]
Comment (irisdv) [reply bot/src/config.rs:61]: yes it starts at 0, the first one is the program name. So if the len = 1 then there is only the program name, if it's greater then we take the second arg at the index = 1

File: bot/src/bot.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+use bigdecimal::{BigDecimal, Zero};
+
+pub fn get_provider(config: &Config) -> SequencerGatewayProvider {
+    if config.devnet_provider.is_devnet {
+        SequencerGatewayProvider::new(
+            Url::from_str(&config.devnet_provider.gateway).unwrap(),
+            Url::from_str(&config.devnet_provider.feeder_gateway).unwrap(),
+        )
+    } else {
+        SequencerGatewayProvider::starknet_alpha_goerli()
```
Comment (Th0rgal) [inline bot/src/bot.rs:36]: what about mainnet?

File: bot/src/bot.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    let find_options = FindOptions::builder().build();
+    let mut domain_cursor = domains
+        .find(
+            doc! { "expiry": { "$lt": BsonDateTime::from_millis(min_expiry_date.timestamp_millis()) } },
+            find_options,
+        )
+        .await?;
+
+    let all_domains = domain_cursor.try_collect::<Vec<_>>().await?;
+    let futures: Vec<_> = all_domains
```
Comment (Th0rgal) [inline bot/src/bot.rs:58]: Doing one query per domain won't scale very well, let's try to do a single query with an aggregate and a project.

File: bot/src/bot.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            let domain_encoded = starknet::id::encode(domain_name)
+                .map_err(|_| anyhow!("Failed to encode domain name"))
+                .context("Error occurred while encoding domain name")?;
+
+            let call_result = provider
+                .call_contract(
+                    CallFunction {
+                        contract_address: config.contract.pricing,
+                        entry_point_selector: selector!("compute_renew_price"),
+                        calldata: vec![domain_encoded, FieldElement::from_dec_str("60").unwrap()],
```
Comment (Th0rgal) [inline bot/src/bot.rs:110]: this felt could be loaded from a const instead of parsed every time

**Review by Th0rgal — CHANGES_REQUESTED** [review]
Please improve the query, use lazy_static or a const when possible wth felts. I think the current version only supports goerli and devnet, would be nice to be able to use mainnet too.

File: bot/src/discord.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
+use anyhow::{Context, Result};
+use serde_json::json;
+use serenity::http::Http;
+use serenity::model::channel::Message;
+use starknet::core::types::FieldElement;
+
+use crate::config::Config;
+
+pub async fn send_message_to_discord(config: &Config, message: &str) -> Result<Message> {
+    let http = Http::new(&config.discord.token);
```
[the remark this answers is not in the document]
Comment (irisdv) [reply bot/src/discord.rs:10]: I used serenity to send messages : https://github.com/serenity-rs/serenity

File: bot/src/bot.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    let mut calldata: Vec<FieldElement> = Vec::new();
+    calldata.push(FieldElement::from_dec_str(&domains.0.len().to_string()).unwrap());
+    calldata.extend_from_slice(&domains.0);
+    calldata.push(FieldElement::from_dec_str(&domains.1.len().to_string()).unwrap());
+    calldata.extend_from_slice(&domains.1);
+
+    let result = account
+        .execute(vec![Call {
+            to: config.contract.renewal,
+            selector: get_selector_from_name("batch_renew").unwrap(),
```
[the remark this answers is not in the document]
Comment (irisdv) [reply bot/src/bot.rs:162]: actually I could use the `selector!` macro no ?

[the remark this answers is not in the document]
Comment (Th0rgal) [reply bot/src/bot.rs]: Yes I didn't know it was ran at compile time, it's much better :)

File: bot/src/discord.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Th0rgal) [reply bot/src/discord.rs:10]: Okay I checked and it's just a wrapper for reqwest::RequestBuilder so no need to cache it :)

**Comment by Th0rgal** [comment]
great work, lgtm

**Pull request merged**